File: apps/web/src/api/skillInventory.ts
import { get } from './client';

export interface SkillInstallation {
  installation_id: string;
  locator_sha256: string;
  environment: { id: string; name: string };
  device: { id: string; identity: string; revoked: boolean };
  presence: 'observed_not_verified_current';
  relationship_status: 'unresolved';
  effective_permissions: null;
  latest_observation: null | {
    observation_id: string; manifest_sha256: string; parser_version: string;
    parse_status: 'parsed' | 'missing_frontmatter' | 'unsupported' | 'invalid_utf8';
    name: string | null; allowed_tools_present: boolean; declared_tools: string[];
    observed_at: string; batch_digest: string;
  };
}
export interface SkillInventoryPage {
  schema_version: 'enterprise-skill-inventory/v1'; items: SkillInstallation[]; next_cursor: string | null;
}
const object = (v: unknown): v is Record<string, unknown> => !!v && typeof v === 'object' && !Array.isArray(v);
const text = (v: unknown): v is string => typeof v === 'string' && v.length > 0 && v.length <= 4096;
const digest = (v: unknown) => typeof v === 'string' && /^[a-f0-9]{64}$/.test(v);
const id = (v: unknown): v is string => typeof v === 'string' && /^ski_[A-Za-z0-9_-]{1,60}$/.test(v);
function installation(v: unknown): v is SkillInstallation {
  if (!object(v) || !id(v.installation_id) || !digest(v.locator_sha256)
    || !object(v.environment) || !text(v.environment.id) || !text(v.environment.name)
    || !object(v.device) || !text(v.device.id) || !text(v.device.identity) || typeof v.device.revoked !== 'boolean'
    || v.presence !== 'observed_not_verified_current' || v.relationship_status !== 'unresolved'
    || v.effective_permissions !== null) return false;
  const o = v.latest_observation;
  if (o === null) return true;
  return isSkillObservation(o);
}
export type SkillObservation = NonNullable<SkillInstallation['latest_observation']>;
export function isSkillObservation(o: unknown): o is SkillObservation {
  if (!object(o) || !text(o.observation_id) || !digest(o.manifest_sha256) || !digest(o.batch_digest)
    || o.parser_version !== 'enterprise-skill-manifest/v1' || typeof o.allowed_tools_present !== 'boolean'
    || !Array.isArray(o.declared_tools) || o.declared_tools.length > 64 || !o.declared_tools.every(text)
    || new Set(o.declared_tools).size !== o.declared_tools.length
    || !text(o.observed_at) || !Number.isFinite(Date.parse(o.observed_at))) return false;
  if (o.parse_status === 'parsed') return text(o.name) && (o.allowed_tools_present || o.declared_tools.length === 0);
  return ['missing_frontmatter', 'unsupported', 'invalid_utf8'].includes(String(o.parse_status))
    && o.name === null && !o.allowed_tools_present && o.declared_tools.length === 0;
}
export function isSkillInventoryPage(v: unknown): v is SkillInventoryPage {
  if (!object(v) || v.schema_version !== 'enterprise-skill-inventory/v1' || !Array.isArray(v.items)
    || v.items.length > 200 || !v.items.every(installation)) return false;
  const ids = v.items.map(row => row.installation_id);
  return ids.every((value, index) => index === 0 || value > ids[index - 1])
    && (v.next_cursor === null || (id(v.next_cursor) && ids.length > 0 && v.next_cursor === ids[ids.length - 1]));
}
export async function getSkillInventory(environmentId: string, deviceId: string, cursor?: string): Promise<SkillInventoryPage> {
  const value = await get<unknown>('/skill-installations', { query: {
    environment_id: environmentId || undefined, device_id: deviceId || undefined, cursor, limit: 50,
  } });
  if (!isSkillInventoryPage(value) || value.items.some(row =>
    (environmentId && row.environment.id !== environmentId) || (deviceId && row.device.id !== deviceId)
    || (cursor && row.installation_id <= cursor))) throw new Error('技能清单响应无法核验');
  return value;
}

File: apps/control-api/app/routers/skill_inventory.py
"""Tenant-scoped historical skill observations, never effective permissions."""

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from sqlalchemy import and_, func, or_, select
from sqlalchemy.orm import Session, aliased

from app.db import get_session
from app.models import EdgeAgent, Environment, SkillInstallation, SkillManifestObservation
from app.security import Identity, ensure_permission, get_identity

router = APIRouter(tags=["inventory"])


def inventory_query(tenant_id: str):
    latest = (
        select(
            SkillManifestObservation,
            func.row_number()
            .over(
                partition_by=SkillManifestObservation.installation_id,
                order_by=(SkillManifestObservation.observed_at.desc(), SkillManifestObservation.id.desc()),
            )
            .label("position"),
        )
        .where(SkillManifestObservation.tenant_id == tenant_id)
        .subquery()
    )
    observation = aliased(SkillManifestObservation, latest)
    return (
        select(SkillInstallation, EdgeAgent, Environment, observation)
        .join(EdgeAgent, EdgeAgent.id == SkillInstallation.edge_agent_id)
        .join(Environment, Environment.id == EdgeAgent.environment_id)
        .outerjoin(observation, (observation.installation_id == SkillInstallation.id) & (latest.c.position == 1))
        .where(SkillInstallation.tenant_id == tenant_id, Environment.tenant_id == tenant_id)
    )


def project_observation(observation):
    if observation is None:
        return None
    return {
        "observation_id": observation.id,
        "manifest_sha256": observation.manifest_sha256,
        "parser_version": observation.parser_version,
        "parse_status": observation.parse_status,
        "name": observation.name,
        "allowed_tools_present": observation.allowed_tools_present,
        "declared_tools": observation.declared_tools,
        "observed_at": observation.observed_at.isoformat() + "Z",
        "batch_digest": observation.batch_digest,
    }


def project(row):
    installation, edge, environment, observation = row
    return {
        "installation_id": installation.id,
        "locator_sha256": installation.locator_sha256,
        "environment": {"id": environment.id, "name": environment.name},
        "device": {"id": edge.id, "identity": edge.device_identity, "revoked": edge.revoked_at is not None},
        "presence": "observed_not_verified_current",
        "relationship_status": "unresolved",
        "effective_permissions": None,
        "latest_observation": project_observation(observation),
    }


@router.get("/api/v1/skill-installations")
def list_skill_installations(
    response: Response,
    environment_id: str | None = Query(default=None, max_length=128),
    device_id: str | None = Query(default=None, max_length=128),
    cursor: str | None = Query(default=None, pattern=r"^ski_[A-Za-z0-9_-]{1,60}$"),
    limit: int = Query(default=50, ge=1, le=200),
    identity: Identity = Depends(get_identity),
    session: Session = Depends(get_session),
):
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    query = inventory_query(identity.tenant_id)
    if environment_id is not None:
        query = query.where(Environment.id == environment_id)
    if device_id is not None:
        query = query.where(EdgeAgent.id == device_id)
    if cursor is not None:
        query = query.where(SkillInstallation.id > cursor)
    rows = session.execute(query.order_by(SkillInstallation.id).limit(limit + 1)).all()
    response.headers["Cache-Control"] = "no-store"
    return {
        "schema_version": "enterprise-skill-inventory/v1",
        "items": [project(row) for row in rows[:limit]],
        "next_cursor": rows[limit - 1][0].id if len(rows) > limit else None,
    }


@router.get("/api/v1/skill-installations/{installation_id}")
def skill_installation_detail(
    installation_id: str,
    response: Response,
    identity: Identity = Depends(get_identity),
    session: Session = Depends(get_session),
):
    row = session.execute(inventory_query(identity.tenant_id).where(SkillInstallation.id == installation_id)).first()
    if row is None:
        raise HTTPException(status_code=404, detail="not_found")
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    response.headers["Cache-Control"] = "no-store"
    return project(row)


@router.get("/api/v1/skill-installations/{installation_id}/observations")
def skill_installation_history(
    installation_id: str,
    response: Response,
    cursor: str | None = Query(default=None, pattern=r"^smo_[A-Za-z0-9_-]{1,60}$"),
    limit: int = Query(default=50, ge=1, le=200),
    identity: Identity = Depends(get_identity),
    session: Session = Depends(get_session),
):
    row = session.execute(inventory_query(identity.tenant_id).where(SkillInstallation.id == installation_id)).first()
    if row is None:
        raise HTTPException(status_code=404, detail="not_found")
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    query = select(SkillManifestObservation).where(
        SkillManifestObservation.tenant_id == identity.tenant_id,
        SkillManifestObservation.installation_id == installation_id,
    )
    if cursor is not None:
        anchor = session.scalar(query.where(SkillManifestObservation.id == cursor))
        if anchor is None:
            raise HTTPException(status_code=404, detail="not_found")
        query = query.where(
            or_(
                SkillManifestObservation.observed_at < anchor.observed_at,
                and_(
                    SkillManifestObservation.observed_at == anchor.observed_at, SkillManifestObservation.id < anchor.id
                ),
            )
        )
    rows = session.scalars(
        query.order_by(
            SkillManifestObservation.observed_at.desc(),
            SkillManifestObservation.id.desc(),
        ).limit(limit + 1)
    ).all()
    response.headers["Cache-Control"] = "no-store"
    return {
        "schema_version": "enterprise-skill-history/v1",
        "installation_id": installation_id,
        "items": [project_observation(row) for row in rows[:limit]],
        "next_cursor": rows[limit - 1].id if len(rows) > limit else None,
    }
